URI path filtering support in cluster stats API

SwethaGuptha opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
Introduced new URI paths  to fetch specific information from _cluster/stats without having to fetch the entire stats that will benefit the API's performance. Motivation from https://github.com/opensearch-project/OpenSearch/pull/14447. The [_cluster/stats API](https://opensearch.org/docs/latest/api-reference/cluster-api/cluster-stats/#response-body-fields) primarily comprises of nodes stats and indices stats, with the new URI paths, properties of these stats objects are flattened and clients can request for specific information.

URI paths introduced:
```
/_cluster/stats/{metric}/nodes/{nodeId}
/_cluster/stats/{metric}/{index_metric}/nodes/{nodeId}

```

**Metrics supported:**
```os, process, jvm, fs, plugins, ingest, network_types, discovery_types, packaging_types, indices, _all```

**Index Metrics supported:**
```shards, docs, store , fielddata, query_cache, completion, segments, analysis, mappings, _all```

**Example REST request formats to exclude the information for**

Request for fetching specific index stats properties only:
```
$ curl localhost:9200/_cluster/stats/indices/docs,segments/nodes/_all                                 
{"_nodes":{"total":1,"successful":1,"failed":0},"cluster_name":"runTask","cluster_uuid":"mtnrtcVURy2hOJn5scMSmQ","timestamp":1728897501917,"status":"green","indices":{"count":0,"docs":{"count":0,"deleted":0},"segments":{"count":0,"memory_in_bytes":0,"terms_memory_in_bytes":0,"stored_fields_memory_in_bytes":0,"term_vectors_memory_in_bytes":0,"norms_memory_in_bytes":0,"points_memory_in_bytes":0,"doc_values_memory_in_bytes":0,"index_writer_memory_in_bytes":0,"version_map_memory_in_bytes":0,"fixed_bit_set_memory_in_bytes":0,"max_unsafe_auto_id_timestamp":-9223372036854775808,"remote_store":{"upload":{"total_upload_size":{"started_bytes":0,"succeeded_bytes":0,"failed_bytes":0},"refresh_size_lag":{"total_bytes":0,"max_bytes":0},"max_refresh_time_lag_in_millis":0,"total_time_spent_in_millis":0,"pressure":{"total_rejections":0}},"download":{"total_download_size":{"started_bytes":0,"succeeded_bytes":0,"failed_bytes":0},"total_time_spent_in_millis":0}},"segment_replication":{"max_bytes_behind":0,"total_bytes_behind":0,"max_replication_lag":0},"file_sizes":{}}}}%  

$ curl localhost:9200/_cluster/stats/indices/mappings,analysis/nodes/_all
{"_nodes":{"total":1,"successful":1,"failed":0},"cluster_name":"runTask","cluster_uuid":"mtnrtcVURy2hOJn5scMSmQ","timestamp":1728897531305,"status":"green","indices":{"count":0,"mappings":{"field_types":[]},"analysis":{"char_filter_types":[],"tokenizer_types":[],"filter_types":[],"analyzer_types":[],"built_in_char_filters":[],"built_in_tokenizers":[],"built_in_filters":[],"built_in_analyzers":[]}}}% 
```
Request for fetching specific node stats properties only:
```
$ curl localhost:9200/_cluster/stats/os,jvm/nodes/_all
{"_nodes":{"total":1,"successful":1,"failed":0},"cluster_name":"runTask","cluster_uuid":"mtnrtcVURy2hOJn5scMSmQ","timestamp":1728897551016,"status":"green","nodes":{"count":{"total":1,"cluster_manager":1,"coordinating_only":0,"data":1,"ingest":1,"master":1,"remote_cluster_client":1,"search":0},"versions":["3.0.0"],"os":{"available_processors":12,"allocated_processors":12,"names":[{"name":"Mac OS X","count":1}],"pretty_names":[{"pretty_name":"Mac OS X","count":1}],"mem":{"total_in_bytes":17179869184,"free_in_bytes":22298624,"used_in_bytes":17157570560,"free_percent":0,"used_percent":100}},"jvm":{"max_uptime_in_millis":495233,"versions":[{"version":"19.0.1","vm_name":"OpenJDK 64-Bit Server VM","vm_version":"19.0.1+10-21","vm_vendor":"Oracle Corporation","bundled_jdk":true,"using_bundled_jdk":false,"count":1}],"mem":{"heap_used_in_bytes":106676736,"heap_max_in_bytes":536870912},"threads":25}}}% 

```

Request for fetching specific node stats and indices state properties:
```
$ curl localhost:9200/_cluster/stats/indices,os,jvm/shards/nodes/_all          
{"_nodes":{"total":1,"successful":1,"failed":0},"cluster_name":"runTask","cluster_uuid":"mtnrtcVURy2hOJn5scMSmQ","timestamp":1728897709009,"status":"yellow","indices":{"count":1,"shards":{"total":3,"primaries":3,"replication":0.0,"index":{"shards":{"min":3,"max":3,"avg":3.0},"primaries":{"min":3,"max":3,"avg":3.0},"replication":{"min":0.0,"max":0.0,"avg":0.0}}}},"nodes":{"count":{"total":1,"cluster_manager":1,"coordinating_only":0,"data":1,"ingest":1,"master":1,"remote_cluster_client":1,"search":0},"versions":["3.0.0"],"os":{"available_processors":12,"allocated_processors":12,"names":[{"name":"Mac OS X","count":1}],"pretty_names":[{"pretty_name":"Mac OS X","count":1}],"mem":{"total_in_bytes":17179869184,"free_in_bytes":27590656,"used_in_bytes":17152278528,"free_percent":0,"used_percent":100}},"jvm":{"max_uptime_in_millis":653256,"versions":[{"version":"19.0.1","vm_name":"OpenJDK 64-Bit Server VM","vm_version":"19.0.1+10-21","vm_vendor":"Oracle Corporation","bundled_jdk":true,"using_bundled_jdk":false,"count":1}],"mem":{"heap_used_in_bytes":141279744,"heap_max_in_bytes":536870912},"threads":31}}}%


$ curl localhost:9200/_cluster/stats/_all/shards/nodes/_all
{"_nodes":{"total":1,"successful":1,"failed":0},"cluster_name":"runTask","cluster_uuid":"mtnrtcVURy2hOJn5scMSmQ","timestamp":1728897740389,"status":"yellow","indices":{"count":1,"shards":{"total":3,"primaries":3,"replication":0.0,"index":{"shards":{"min":3,"max":3,"avg":3.0},"primaries":{"min":3,"max":3,"avg":3.0},"replication":{"min":0.0,"max":0.0,"avg":0.0}}}},"nodes":{"count":{"total":1,"cluster_manager":1,"coordinating_only":0,"data":1,"ingest":1,"master":1,"remote_cluster_client":1,"search":0},"versions":["3.0.0"],"os":{"available_processors":12,"allocated_processors":12,"names":[{"name":"Mac OS X","count":1}],"pretty_names":[{"pretty_name":"Mac OS X","count":1}],"mem":{"total_in_bytes":17179869184,"free_in_bytes":260648960,"used_in_bytes":16919220224,"free_percent":2,"used_percent":98}},"process":{"cpu":{"percent":0},"open_file_descriptors":{"min":191,"max":191,"avg":191}},"jvm":{"max_uptime_in_millis":684642,"versions":[{"version":"19.0.1","vm_name":"OpenJDK 64-Bit Server VM","vm_version":"19.0.1+10-21","vm_vendor":"Oracle Corporation","bundled_jdk":true,"using_bundled_jdk":false,"count":1}],"mem":{"heap_used_in_bytes":147571200,"heap_max_in_bytes":536870912},"threads":35},"fs":{"total_in_bytes":499963174912,"free_in_bytes":94558470144,"available_in_bytes":94558470144,"cache_reserved_in_bytes":0},"plugins":[],"network_types":{"transport_types":{"netty4":1},"http_types":{"netty4":1}},"discovery_types":{"zen":1},"packaging_types":[{"type":"tar","count":1}],"ingest":{"number_of_pipelines":0,"processor_stats":{}}}}
```

If a random metric is requested, exception is thrown:

```
$ curl localhost:9200/_cluster/stats/os,mappings,jvm,analysis,addssf/nodes/_all
{"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"request [/_cluster/stats/os,mappings,jvm,analysis,addssf/nodes/_all] contains unrecognized metrics: [addssf], [analysis], [mappings]"}],"type":"illegal_argument_exception","reason":"request [/_cluster/stats/os,mappings,jvm,analysis,addssf/nodes/_all] contains unrecognized metrics: [addssf], [analysis], [mappings]"},"status":400}%
```

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]
<!-- List any other related issues here -->

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

sandeshkr419 commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsResponse.java:
```diff
@@ -105,6 +107,36 @@ public ClusterStatsResponse(
         this.status = status;
     }
 
+    public ClusterStatsResponse(
```
I think with a little refactoring you should be able to invoke this constructor from the above constructor at L67 and reduce some code duplicacy.

SwethaGuptha replied:
Done.

sandeshkr419 commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIndices.java:
```diff
@@ -186,13 +234,27 @@ static final class Fields {
-        completion.toXContent(builder, params);
-        segments.toXContent(builder, params);
+        if (shards != null) {
+            shards.toXContent(builder, params);
```
Instead of adding null checks at all places, why not just add a parameter in the method itself to do nothing if null is received?

SwethaGuptha replied:
Moving the null checks inside the method will cause null pointer exceptions when the objects are null on calling toXContent methods.

sandeshkr419 replied:
@SwethaGuptha Did you get a chance to check this?

rajiv-kv commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsRequest.java:
```diff
@@ -73,12 +83,124 @@ public void useAggregatedNodeLevelResponses(boolean useAggregatedNodeLevelRespon
+        ANALYSIS("analysis", Metric.INDICES),
+        MAPPINGS("mappings", Metric.INDICES),
+
+        OS("os", Metric.NODES),
+        PROCESS("process", Metric.NODES),
+        JVM("jvm", Metric.NODES),
+        FS("fs", Metric.NODES),
```
The nodes stats define these as Metrics although this is reported per-node. Should we keep it consistent. The 2nd metric path is defined to be a `index-metric`
https://opensearch.org/docs/latest/api-reference/nodes-apis/nodes-stats/#path-parameters

rajiv-kv replied:
Discussed offline on this and in this PR we will have the metric filters only along with node filter.
The URI would be `/_cluster/stats/nodes/{nodeId}/{metric}/{index_metric}`

rajiv-kv commented on server/src/main/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsAction.java:
```diff
@@ -54,7 +64,36 @@ public class RestClusterStatsAction extends BaseRestHandler {
+        );
+    }
+
+    static final Map<String, Set<String>> METRIC_TO_SUB_METRICS_MAP;
```
Can the metric/sub-metric map move to `ClusterStatsRequest` class ? Lets initialize all of them from a static method of Metrics class `Metrics#toSubMetricMap`

SwethaGuptha replied:
The METRIC_TO_SUB_METRICS_MAP will not change through-out the process life, wouldn't it be better to have this initialized in the static block instead of invoking the static method.

rajiv-kv replied:
Yes, but can it be intiialized from a static method and the map be part of Metrics class ?

rajiv-kv commented on server/src/main/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsAction.java:
```diff
@@ -64,9 +103,78 @@ public String getName() {
         clusterStatsRequest.timeout(request.param("timeout"));
         clusterStatsRequest.useAggregatedNodeLevelResponses(true);
+
+        if (metrics.size() > 1 & metrics.contains("_all")) {
```
& -> &&

SwethaGuptha replied:
Fixed.

SwethaGuptha commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsRequest.java:
```diff
@@ -48,11 +52,17 @@
 @PublicApi(since = "1.0.0")
 public class ClusterStatsRequest extends BaseNodesRequest<ClusterStatsRequest> {
 
+    private final Set<String> requestedMetrics = new HashSet<>(Metric.allMetrics());
```
Yes.

SwethaGuptha commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIndices.java:
```diff
@@ -68,14 +71,57 @@ public class ClusterStatsIndices implements ToXContentFragment {
+    ) {
         Map<String, ShardStats> countsPerIndex = new HashMap<>();
 
-        this.docs = new DocsStats();
-        this.store = new StoreStats();
-        this.fieldData = new FieldDataStats();
-        this.queryCache = new QueryCacheStats();
-        this.completion = new CompletionStats();
-        this.segments = new SegmentsStats();
```
Yes. All stats are included by default.

Pranshu-S commented on server/src/main/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsAction.java:
```diff
@@ -64,9 +100,95 @@ public String getName() {
-        ClusterStatsRequest clusterStatsRequest = new ClusterStatsRequest().nodesIds(request.paramAsStringArray("nodeId", null));
+        String[] nodeIds = request.paramAsStringArray("nodeId", null);
+        Set<String> metrics = Strings.tokenizeByCommaToSet(request.param("metric", "_all"));
+        Set<String> indexMetrics = Strings.tokenizeByCommaToSet(request.param("index_metric", null));
```
Can we make the default for indexMetrics as `_all`?

SwethaGuptha replied:
This is null to handle cases when the client request doesn't include indices metrics.

rajiv-kv commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsResponse.java:
```diff
@@ -88,12 +90,42 @@ public ClusterStatsResponse(
         this.timestamp = timestamp;
-        nodesStats = new ClusterStatsNodes(nodes);
-        indicesStats = new ClusterStatsIndices(nodes, MappingStats.of(state), AnalysisStats.of(state));
+        nodesStats = !requestedMetrics.isEmpty() && !requestedMetrics.equals(Set.of(Metric.INDICES.metricName()))
```
Is there an assumption that anything other than `INDICES` is nodeStats ? 
If so, can we return empty ClusterStatsNodes instead of null ?

SwethaGuptha replied:
>Is there an assumption that anything other than INDICES is nodeStats ?
yes.

> If so, can we return empty ClusterStatsNodes instead of null ?
I think its okay to return null otherwise we will need to create a non-parameterized constructor for ClusterStatsNodes, metrics classes.

rajiv-kv replied:
If we add new metrics in future that is not related to Nodes, this assumption will break.

SwethaGuptha replied:
Addressed this by maintaining a set of metrics of node stats.

SwethaGuptha commented on server/src/main/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsAction.java:
```diff
@@ -64,9 +100,95 @@ public String getName() {
+                )
+            );
+        } else {
+            clusterStatsRequest.clearRequestedMetrics();
+            clusterStatsRequest.clearIndicesMetrics();
```
To have the change backward compatible, by default the requested metrics and indices metrics are set to all metrics in the request hence these metrics are cleared for computing the specific metrics requested in the rest call.

SwethaGuptha replied:
Makes sense, will introduce a boolean flag to determine if metric filtering needs to be applied or not. By default this flag will be false that will be inferred as compute all metrics at transport layer.

SwethaGuptha commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIndices.java:
```diff
@@ -186,13 +233,27 @@ static final class Fields {
+        if (shards != null) {
+            shards.toXContent(builder, params);
+        }
+        if (docs != null) {
```
I think we should rely on test cases to validate the correctness of the response returned instead of coupling the null check verification with the requested metrics.

rajiv-kv replied:
Response needs to be constructed based on the user request and not based on transport response. It is better to throw errors when the expectation is not met rather than ignore silently.

SwethaGuptha replied:
I agree its good to have those validation but the transport response is constructed based on user request. If a field is requested then the field will be set otherwise it won't be set, so I am not sure if we should validate this here again.

rajiv-kv replied:
If there is a bug in TransportAction and the response is not returned we would never catch the error.  We should assert the expectations of transport response at Rest layer and have it validated by IT .

SwethaGuptha replied:
Will create a follow up git hub issue to address this.

SwethaGuptha commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIndices.java:
```diff
@@ -68,14 +71,56 @@ public class ClusterStatsIndices implements ToXContentFragment {
-        this.segments = new SegmentsStats();
+        Consumer<QueryCacheStats> queryCacheStatsConsumer = (queryCacheStats) -> {
+            if (IndexMetrics.QUERY_CACHE.containedIn(indicesMetrics)) {
+                if (this.queryCache == null) this.queryCache = new QueryCacheStats();
```
This in constructor and the objects are not initialized by default. With the metrics change object will be initialized when the metric is requested.

SwethaGuptha commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/TransportClusterStatsAction.java:
```diff
@@ -177,33 +182,35 @@ protected ClusterStatsNodeResponse nodeOperation(ClusterStatsNodeRequest nodeReq
+                            seqNoStats = null;
+                            retentionLeaseStats = null;
+                        }
+                        shardsStats.add(
```
No, we use the same object to construct other index metrics as well.

rajiv-kv replied:
I see the shardsStats being read conditionally in ClusterStatsIndices based on following condition
` IndexMetrics.SHARDS.containedIn(indicesMetrics)` . Should that also change ?

SwethaGuptha replied:
The ShardStats created here is different from metric SHARDS. The metric ShardStats object  https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIndices.java#L211 holds information only the total number of shard and primary number of shards for an index.

rajiv-kv replied:
Got it. Can we add a test to verify if doc stats are correctly returned with MetricFilter != SHARDS ?

SwethaGuptha commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/TransportClusterStatsAction.java:
```diff
@@ -147,19 +151,20 @@ protected ClusterStatsNodeResponse newNodeResponse(StreamInput in) throws IOExce
-            true,
-            true,
-            true,
+            Metric.OS.containedIn(requestedMetrics),
+            Metric.PROCESS.containedIn(requestedMetrics),
+            Metric.JVM.containedIn(requestedMetrics),
```
Within a class these checks are made only once per metric hence didn't introduce boolean flags.

SwethaGuptha commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIndices.java:
```diff
@@ -92,12 +137,12 @@ public ClusterStatsIndices(List<ClusterStatsNodeResponse> nodeResponses, Mapping
-                queryCache.add(r.getAggregatedNodeLevelStats().commonStats.queryCache);
-                completion.add(r.getAggregatedNodeLevelStats().commonStats.completion);
-                segments.add(r.getAggregatedNodeLevelStats().commonStats.segments);
+                docsStatsConsumer.accept(r.getAggregatedNodeLevelStats().commonStats.docs);
```
The fields won't be set before this as they are initialized only part of the constructor.

SwethaGuptha commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsRequestBuilder.java:
```diff
@@ -55,4 +57,16 @@ public final ClusterStatsRequestBuilder useAggregatedNodeLevelResponses(boolean
     }
+
+    public final ClusterStatsRequestBuilder requestMetrics(Set<String> requestMetrics) {
+        request.clearRequestedMetrics();
```
Yes, expectation is that the caller can use this interface to provide list of requested metrics only once. If we want to support incremental change we can expose addMetrics. Thoughts?

SwethaGuptha replied:
Not relevant after https://github.com/opensearch-project/OpenSearch/pull/15938#discussion_r1775621606

SwethaGuptha commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/TransportClusterStatsAction.java:
```diff
@@ -177,33 +182,35 @@ protected ClusterStatsNodeResponse nodeOperation(ClusterStatsNodeRequest nodeReq
+                            new ShardStats(
+                                indexShard.routingEntry(),
+                                indexShard.shardPath(),
+                                new CommonStats(indicesService.getIndicesQueryCache(), indexShard, SHARD_STATS_FLAGS),
```
Ack.

SwethaGuptha commented on server/src/main/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsAction.java:
```diff
@@ -64,9 +100,95 @@ public String getName() {
+        String[] nodeIds = request.paramAsStringArray("nodeId", null);
+        Set<String> metrics = Strings.tokenizeByCommaToSet(request.param("metric", "_all"));
+        Set<String> indexMetrics = Strings.tokenizeByCommaToSet(request.param("index_metric", null));
+        logger.info("Request URI path params, metrics : {}, index_metrics : {}, nodes : {}", metrics, indexMetrics, nodeIds);
```
No, will remove this.

rajiv-kv commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsRequest.java:
```diff
@@ -48,11 +52,20 @@
+        if (in.getVersion().onOrAfter(Version.V_3_0_0)) {
+            requestedMetrics.clear();
+            indicesMetrics.clear();
+            requestedMetrics.addAll(in.readStringList());
+            indicesMetrics.addAll(in.readStringList());
```
Yes. Can you check the implementation of CommonStatsFlags#Flags ?

SwethaGuptha replied:
Looked at it, I see the idea there is to encode the flags to a long number by bit manipulation on the flag index. I don't feel confident currently making the same change here in terms of correctness and collisions during decoding.

shwetathareja commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsRequest.java:
```diff
@@ -73,12 +86,127 @@ public void useAggregatedNodeLevelResponses(boolean useAggregatedNodeLevelRespon
+     * An enumeration of the "core" sections of indices metrics that may be requested
+     * from the cluster stats endpoint.
+     */
+    public enum IndexMetrics {
```
@PublicApi(since = "3.0.0")

shwetathareja commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsResponse.java:
```diff
@@ -88,12 +90,42 @@ public ClusterStatsResponse(
         this.timestamp = timestamp;
-        nodesStats = new ClusterStatsNodes(nodes);
-        indicesStats = new ClusterStatsIndices(nodes, MappingStats.of(state), AnalysisStats.of(state));
+        nodesStats = !requestedMetrics.isEmpty() && !requestedMetrics.equals(Set.of(Metric.INDICES.metricName()))
```
this is redundant - !requestedMetrics.isEmpty()

SwethaGuptha replied:
Tracking this as part of https://github.com/opensearch-project/OpenSearch/pull/15938#discussion_r1775193908

shwetathareja commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsResponse.java:
```diff
@@ -131,8 +163,13 @@ public void writeTo(StreamOutput out) throws IOException {
+            out.writeOptionalWriteable(indicesStats.getMappings());
+            out.writeOptionalWriteable(indicesStats.getAnalysis());
+        } else {
+            out.writeOptionalWriteable(null);
+            out.writeOptionalWriteable(null);
```
did you test, this is working as expected and not causing serialization issue?

SwethaGuptha replied:
yes, verified this.

shwetathareja commented on server/src/main/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsAction.java:
```diff
@@ -54,7 +63,34 @@ public class RestClusterStatsAction extends BaseRestHandler {
+            asList(
+                new Route(GET, "/_cluster/stats"),
+                new Route(GET, "/_cluster/stats/nodes/{nodeId}"),
+                new Route(GET, "/_cluster/stats/{metric}/nodes/{nodeId}"),
+                new Route(GET, "/_cluster/stats/{metric}/{index_metric}/nodes/{nodeId}")
```
it should be nodes/nodeId/{metric} right, same for the other one as well?

SwethaGuptha replied:
From https://github.com/opensearch-project/OpenSearch/pull/15938#discussion_r1766842520, we decided to have path as  /stats/{metric}/nodes instead on /nodes/nodeId/{metric}, so that /stats/ is not a dummy path.

shwetathareja commented on server/src/main/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsAction.java:
```diff
@@ -64,9 +100,95 @@ public String getName() {
+
+            logger.info("Computed metrics: {}", metricsRequested);
+            if (ClusterStatsRequest.Metric.INDICES.containedIn(metricsRequested)) {
+                logger.info("Stats request contains indices metric");
```
not needed

SwethaGuptha commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsRequest.java:
```diff
@@ -48,11 +52,20 @@
             useAggregatedNodeLevelResponses = in.readOptionalBoolean();
         }
+        if (in.getVersion().onOrAfter(Version.V_3_0_0)) {
+            requestedMetrics.clear();
+            indicesMetrics.clear();
```
This will not be needed after https://github.com/opensearch-project/OpenSearch/pull/15938#discussion_r1775621606

SwethaGuptha commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsRequest.java:
```diff
@@ -73,12 +86,127 @@ public void useAggregatedNodeLevelResponses(boolean useAggregatedNodeLevelRespon
+        return new HashSet<>(indicesMetrics);
+    }
+
+    public void clearRequestedMetrics() {
+        requestedMetrics.clear();
+    }
+
+    public void clearIndicesMetrics() {
+        indicesMetrics.clear();
+    }
```
Same as above, this will not be needed after https://github.com/opensearch-project/OpenSearch/pull/15938#discussion_r1775621606

rajiv-kv commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/TransportClusterStatsAction.java:
```diff
@@ -124,14 +134,27 @@ protected ClusterStatsResponse newResponse(
+                responses,
+                failures,
+                state,
+                request.requestedMetrics(),
+                request.indicesMetrics()
```
Why are we sending the request back in response ? This seems to be unnecessary ser/deser

SwethaGuptha replied:
These metrics are not sent back, they are only used in the response constructor to check if node/indices stats need to be computed or not.

rajiv-kv commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsRequest.java:
```diff
@@ -48,11 +52,20 @@
         }
+        if (in.getVersion().onOrAfter(Version.V_3_0_0)) {
+            applyMetricFiltering = in.readOptionalBoolean();
+            requestedMetrics.addAll(in.readStringList());
+            indexMetricsRequested.addAll(in.readStringList());
```
Can we serialize it as boolean array instead of string list

rajiv-kv commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsResponse.java:
```diff
@@ -88,12 +90,42 @@ public ClusterStatsResponse(
+            nodes,
+            failures,
+            state,
+            ClusterStatsRequest.Metric.allMetrics(),
+            ClusterStatsRequest.IndexMetrics.allIndicesMetrics()
```
Is this required ?

SwethaGuptha replied:
Yes, this change was done to avoid duplication of code.

rajiv-kv commented on server/src/internalClusterTest/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIT.java:
```diff
@@ -505,6 +507,124 @@ public void testNodeRolesWithDataNodeLegacySettings() throws ExecutionException,
+        assertNotNull(response.getNodesStats().getJvm());
+        assertNotNull(response.getNodesStats().getOs());
+
+        response = client().admin()
```
Lets seperate it out as three tests and assert on the values also.

SwethaGuptha replied:
Done.

rajiv-kv commented on server/src/internalClusterTest/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIT.java:
```diff
@@ -505,6 +507,124 @@ public void testNodeRolesWithDataNodeLegacySettings() throws ExecutionException,
         assertEquals(expectedNodesRoles, Set.of(getNodeRoles(client, 0), getNodeRoles(client, 1)));
     }
 
+    public void testClusterStatsMetricFiltering() {
```
Could we also parametrize the existing tests to turn on Transport level metrics filter ? You can mention all the metrics in the filter.

SwethaGuptha replied:
The state of the cluster is retained within an IT, this will lead to conflicts on parametrizing within an IT.

rajiv-kv commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/TransportClusterStatsAction.java:
```diff
@@ -177,33 +202,41 @@ protected ClusterStatsNodeResponse nodeOperation(ClusterStatsNodeRequest nodeReq
+        if (!applyMetricFiltering || Metric.INDICES.containedIn(requestedMetrics)) {
+            CommonStatsFlags commonStatsFlags = new CommonStatsFlags();
+            for (String metric : nodeRequest.request.indicesMetrics()) {
+                if (INDEX_METRIC_TO_SHARDS_STATS_FLAG_MAP.containsKey(metric)) {
```
Should we throw error if it does not contain ?

SwethaGuptha replied:
All index metrics are not common stats, so can't throw an error here.

rajiv-kv commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/TransportClusterStatsAction.java:
```diff
@@ -147,19 +170,21 @@ protected ClusterStatsNodeResponse newNodeResponse(StreamInput in) throws IOExce
-            true,
-            true,
-            true,
+            !applyMetricFiltering || Metric.OS.containedIn(requestedMetrics),
```
Can we introduce a private method `statsRequired(request, Metric)`

rajiv-kv commented on server/src/main/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsAction.java:
```diff
@@ -64,9 +100,95 @@ public String getName() {
+        Set<String> indexMetrics = Strings.tokenizeByCommaToSet(request.param("index_metric", null));
+        logger.info("Request URI path params, metrics : {}, index_metrics : {}, nodes : {}", metrics, indexMetrics, nodeIds);
+
+        ClusterStatsRequest clusterStatsRequest = new ClusterStatsRequest().nodesIds(nodeIds);
         clusterStatsRequest.timeout(request.param("timeout"));
         clusterStatsRequest.useAggregatedNodeLevelResponses(true);
```
nit: `ClusterStatsRequest` object can be constructed after metrics validation

rajiv-kv commented on server/src/main/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsAction.java:
```diff
@@ -64,9 +100,93 @@ public String getName() {
+                metricsRequested.addAll(metrics);
+            }
+            final Set<String> invalidMetrics = new TreeSet<>();
+            for (String metric : metricsRequested) {
```
Do we need to add tests for invalid metrics/index_metrics validation ?

SwethaGuptha replied:
Added.

rajiv-kv commented on server/src/test/java/org/opensearch/action/admin/cluster/stats/ClusterStatsRequestBuilderTests.java:
```diff
@@ -0,0 +1,80 @@
+        clusterStatsRequestBuilder.requestMetrics(Set.of(Metric.INDICES.metricName(), Metric.JVM.metricName()));
+        clusterStatsRequestBuilder.indexMetrics(Set.of(IndexMetrics.MAPPINGS.metricName(), IndexMetrics.ANALYSIS.metricName()));
+        assertEquals(Set.of(Metric.INDICES.metricName(), Metric.JVM.metricName()), clusterStatsRequestBuilder.request().requestedMetrics());
+        assertEquals(
```
Can we also assert on the boolean `applyMetricFiltering`

rajiv-kv commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/TransportClusterStatsAction.java:
```diff
@@ -223,6 +255,10 @@ protected ClusterStatsNodeResponse nodeOperation(ClusterStatsNodeRequest nodeReq
     }
 
+    private boolean isMetricRequired(boolean applyMetricFilter, Metric metric, Set<Metric> requestedMetrics) {
+        return !applyMetricFilter || requestedMetrics.contains(metric);
```
should this be `applyMetricFilter && requestedMetrics.contains(metric)`

SwethaGuptha replied:
We want to compute all metrics when the filter is not enabled hence using the OR boolean condition.

rajiv-kv commented on server/src/main/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsAction.java:
```diff
@@ -64,9 +100,93 @@ public String getName() {
+
+            final Set<String> metricsRequested = new HashSet<>();
+            if (metrics.contains("_all")) {
+                metricsRequested.addAll(METRIC_REQUEST_CONSUMER_MAP.keySet());
```
should we add `indexMetricsRequested.addAll(INDEX_METRIC_TO_REQUEST_CONSUMER_MAP.keySet());` as well here ?

rajiv-kv replied:
Yes populate the subset in that case.

rajiv-kv commented on server/src/main/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsAction.java:
```diff
@@ -64,9 +100,93 @@ public String getName() {
+            } else {
+                metricsRequested.addAll(metrics);
+            }
+            final Set<String> invalidMetrics = new TreeSet<>();
```
Is the validation required only for the else block of line 139 ?

SwethaGuptha replied:
Same as https://github.com/opensearch-project/OpenSearch/pull/15938#discussion_r1780647863

rajiv-kv commented on server/src/main/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsAction.java:
```diff
@@ -64,9 +100,93 @@ public String getName() {
+                } else {
+                    indexMetricsRequested.addAll(indexMetrics);
+                }
+                final Set<String> invalidIndexMetrics = new TreeSet<>();
```
validation required only if `indexMetricsRequested` is not empty

SwethaGuptha replied:
This will introducing more branching in the code , making it difficult to understand the filtering logic.

rajiv-kv replied:
If the default values are set by the code and not coming from request, it is un-necessary to validate them. Would an early branching help ?

rajiv-kv commented on server/src/test/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsActionTests.java:
```diff
@@ -0,0 +1,85 @@
+import static org.hamcrest.object.HasToString.hasToString;
+import static org.mockito.Mockito.mock;
+
+public class RestClusterStatsActionTests extends OpenSearchTestCase {
```
Can we have one success case ?

rajiv-kv replied:
OK. I think with properly inserted mocks you should be able to verify. Lets create a follow-up PR for this. Can you add the manual testing details to PR ?

SwethaGuptha replied:
Extracted the logic for cluster stats request computation to a static block to make it visible for testing.

rajiv-kv commented on server/src/test/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsActionTests.java:
```diff
@@ -0,0 +1,85 @@
+        final HashMap<String, String> params = new HashMap<>();
+        final String metric = randomAlphaOfLength(64);
+        params.put("metric", metric);
+        final RestRequest request = new FakeRestRequest.Builder(xContentRegistry()).withPath("/_cluster/stats").withParams(params).build();
```
Can we have the metric and index_metric as part of Path and not Params ?

SwethaGuptha replied:
FakeRestRequest doesn't have support to compute the params from URI path. Responsibility is with the clients to set params.

rajiv-kv commented on server/src/test/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsActionTests.java:
```diff
@@ -0,0 +1,85 @@
+        assertThat(e, hasToString(containsString("request [/_cluster/stats] contains unrecognized metric: [" + metric + "]")));
+    }
+
+    public void testUnrecognizedIndexMetric() {
```
Could we test empty `index_metric` path?
`/_cluster/stats/_all//nodes/_all` ?

SwethaGuptha replied:
This is a valid case, the tests in these files are for failure cases.

rajiv-kv replied:
What is the expected output in this case ?

SwethaGuptha replied:
This will be treated as /_cluster/stats/_all/_all/nodes/_all, hence all stats will be returned.

SwethaGuptha commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/TransportClusterStatsAction.java:
```diff
@@ -177,33 +201,41 @@ protected ClusterStatsNodeResponse nodeOperation(ClusterStatsNodeRequest nodeReq
+        if (isMetricRequired(applyMetricFiltering, Metric.INDICES, requestedMetrics)) {
+            CommonStatsFlags commonStatsFlags = new CommonStatsFlags();
+            for (ClusterStatsRequest.IndexMetric indexMetric : nodeRequest.request.indicesMetrics()) {
+                if (INDEX_METRIC_TO_SHARDS_STATS_FLAG_MAP.containsKey(indexMetric)) {
```
Can't throw an exception here as all metric of IndexMetrics are not part of common stats flags.

SwethaGuptha replied:
Reversed the map to store key as commons stats flag and value as index metric .

rajiv-kv commented on server/src/test/java/org/opensearch/action/admin/cluster/stats/ClusterStatsRequestBuilderTests.java:
```diff
@@ -0,0 +1,81 @@
+            this.testClient,
+            ClusterStatsAction.INSTANCE
+        );
+        clusterStatsRequestBuilder.useAggregatedNodeLevelResponses(false);
+        assertFalse(clusterStatsRequestBuilder.request().useAggregatedNodeLevelResponses());
```
Can we serialize and de-serialize the request before asserting for the values ?

SwethaGuptha replied:
Serde test cases for ClusterStatsRequest are written separately, I don't think we need to cover serde here.

rajiv-kv replied:
Ok. Do you think we can combine with them by constructing the request via builder ? Not seeing much value in having a test-class for validation of setters.

SwethaGuptha replied:
Makes sense, it should be okay to delete the builder test file.

rajiv-kv commented on server/src/test/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsActionTests.java:
```diff
@@ -0,0 +1,159 @@
+        assertNotNull(clusterStatsRequest);
+        assertTrue(clusterStatsRequest.useAggregatedNodeLevelResponses());
+        assertTrue(clusterStatsRequest.computeAllMetrics());
+        assertFalse(clusterStatsRequest.requestedMetrics().isEmpty());
+        assertEquals(3, clusterStatsRequest.requestedMetrics().size());
+        for (ClusterStatsRequest.Metric metric : metricsRequested) {
+            assertTrue(clusterStatsRequest.requestedMetrics().contains(metric));
+        }
```
We can rather have local `metricsRequested ` as list and perform  List#equals comparison

SwethaGuptha replied:
Thanks. Updated.

SwethaGuptha commented on server/src/main/java/org/opensearch/index/cache/query/QueryCacheStats.java:
```diff
@@ -77,6 +77,9 @@ public QueryCacheStats(long ramBytesUsed, long hitCount, long missCount, long ca
     }
 
     public void add(QueryCacheStats stats) {
+        if (stats == null) {
+            return;
+        }
```
This can be null when query cache stats are not requested.  There were null pointer exception when computing AggregatedNodeLevelStats. For other stats, null check is already present.

rajiv-kv replied:
Lets add the null check at the caller only

SwethaGuptha commented on server/src/test/java/org/opensearch/action/admin/cluster/stats/ClusterStatsRequestTests.java:
```diff
@@ -0,0 +1,49 @@
+        Version testVersion = VersionUtils.randomVersionBetween(random(), Version.CURRENT.minimumCompatibilityVersion(), Version.CURRENT);
+
+        BytesStreamOutput output = new BytesStreamOutput();
+        output.setVersion(testVersion);
```
The version is dynamically selected, it can be a current version or any version before that.

SwethaGuptha commented on server/src/test/java/org/opensearch/action/admin/cluster/stats/ClusterStatsRequestTests.java:
```diff
@@ -0,0 +1,49 @@
+        streamInput.setVersion(testVersion);
+        ClusterStatsRequest deserializedClusterStatsRequest = new ClusterStatsRequest(streamInput);
+
+        assertEquals(clusterStatsRequest.computeAllMetrics(), deserializedClusterStatsRequest.computeAllMetrics());
+        assertEquals(clusterStatsRequest.requestedMetrics(), deserializedClusterStatsRequest.requestedMetrics());
+        assertEquals(clusterStatsRequest.indicesMetrics(), deserializedClusterStatsRequest.indicesMetrics());
+        assertEquals(
```
For current and versions after it is expected to be always set.

SwethaGuptha commented on server/src/main/java/org/opensearch/index/fielddata/FieldDataStats.java:
```diff
@@ -80,6 +80,9 @@ public FieldDataStats(long memorySize, long evictions, @Nullable FieldMemoryStat
     }
 
     public void add(FieldDataStats stats) {
+        if (stats == null) {
+            return;
+        }
```
Same as https://github.com/opensearch-project/OpenSearch/pull/15938#discussion_r1787036376

SwethaGuptha commented on server/src/main/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsAction.java:
```diff
@@ -64,10 +100,104 @@ public String getName() {
+            }
+
+            Set<String> indexMetrics = Strings.tokenizeByCommaToSet(
+                request.param("index_metric", metricsRequested.contains(ClusterStatsRequest.Metric.INDICES.metricName()) ? "_all" : null)
```
Do you mean local variable for index_metric

SwethaGuptha commented on server/src/internalClusterTest/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIT.java:
```diff
@@ -317,6 +321,52 @@ public void testValuesSmokeScreen() throws IOException, ExecutionException, Inte
+            .prepareClusterStats()
+            .useAggregatedNodeLevelResponses(randomBoolean())
+            .computeAllMetrics(false)
+            .requestMetrics(Set.of(Metric.values()))
+            .indexMetrics(Set.of(IndexMetric.values()))
```
This test case is a variation of existing test case https://github.com/SwethaGuptha/OpenSearch/blob/main/server/src/internalClusterTest/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIT.java#L277 with URI path filtering enabled. Test cases for different combinations are covered separately.

SwethaGuptha commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIndices.java:
```diff
@@ -68,14 +71,56 @@ public class ClusterStatsIndices implements ToXContentFragment {
     private MappingStats mappings;
 
     public ClusterStatsIndices(List<ClusterStatsNodeResponse> nodeResponses, MappingStats mappingStats, AnalysisStats analysisStats) {
+        this(IndexMetrics.allIndicesMetrics(), nodeResponses, mappingStats, analysisStats);
```
I will prefer to not combine it with this PR and address it in a separate PR. This PR has other smaller necessary changes will increase the diff for the reviewer.

SwethaGuptha commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsRequest.java:
```diff
@@ -48,11 +50,30 @@
 
+    private final Set<Metric> requestedMetrics = new HashSet<>();
+    private final Set<IndexMetric> indexMetricsRequested = new HashSet<>();
+    private Boolean computeAllMetric = true;
```
As discuss will take this as a follow-up.

SwethaGuptha commented on server/src/test/java/org/opensearch/action/admin/cluster/stats/ClusterStatsRequestTests.java:
```diff
@@ -0,0 +1,49 @@
+        clusterStatsRequest.writeTo(output);
+
+        StreamInput streamInput = output.bytes().streamInput();
+        streamInput.setVersion(testVersion);
```
Mixed version serialisation/deserialisation will not represent real scenario here. For Serialisation with 3.x and deserialisation with 2.x, the byte buffers will not be full read and no exception is thrown. On an actual domain the process throws an exception like "buffer not fully read".

rajiv-kv replied:
Thanks for adding test for 2.17 and 3.0.0. Lets's test mixed version manually for now.

SwethaGuptha commented on server/src/test/java/org/opensearch/action/admin/cluster/stats/ClusterStatsResponseTests.java:
```diff
@@ -0,0 +1,231 @@
+            defaultClusterStatsNodeResponses,
+            List.of(),
+            ClusterState.EMPTY_STATE,
+            Set.of(),
+            Set.of()
```
Added.

rajiv-kv commented on server/src/main/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsAction.java:
```diff
@@ -64,10 +102,102 @@ public String getName() {
+
+    public static ClusterStatsRequest fromRequest(final RestRequest request) {
+        Set<String> metrics = Strings.tokenizeByCommaToSet(request.param("metric", "_all"));
+        String indicesMetricsDefaultValue = metrics.contains(Metric.INDICES.metricName()) || metrics.contains("_all") ? "_all" : null;
```
can you add a code comment and as well to the documentaion that default filter is _all for indicesMetrics ?

SwethaGuptha replied:
Added.

rajiv-kv commented on server/src/main/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsAction.java:
```diff
@@ -64,10 +102,102 @@ public String getName() {
+                    );
+                }
+            }
+            clusterStatsRequest.computeAllMetrics(false);
```
should this be outside the if block and always set to `false` ?

rajiv-kv commented on server/src/main/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsAction.java:
```diff
@@ -64,10 +102,102 @@ public String getName() {
         clusterStatsRequest.useAggregatedNodeLevelResponses(true);
-        return channel -> client.admin().cluster().clusterStats(clusterStatsRequest, new NodesResponseRestListener<>(channel));
+
+        if (!metrics.isEmpty()) {
```
can this be ever empty ?

SwethaGuptha replied:
No, removed the if check.

rajiv-kv commented:
Please take a look at these failing tests in build
```
[org.opensearch.search.nested.SimpleNestedIT.testCheckFixedBitSetCache {p0={"search.concurrent_segment_search.enabled":"true"}}](https://build.ci.opensearch.org/job/gradle-check/48900/testReport/junit/org.opensearch.search.nested/SimpleNestedIT/testCheckFixedBitSetCache__p0___search_concurrent_segment_search_enabled___true___/)
[org.opensearch.search.nested.SimpleNestedIT.testCheckFixedBitSetCache {p0={"search.concurrent_segment_search.enabled":"false"}}](https://build.ci.opensearch.org/job/gradle-check/48900/testReport/junit/org.opensearch.search.nested/SimpleNestedIT/testCheckFixedBitSetCache__p0___search_concurrent_segment_search_enabled___false___/)
```

SwethaGuptha commented:
> Please take a look at these failing tests in build
> 
> ```
> [org.opensearch.search.nested.SimpleNestedIT.testCheckFixedBitSetCache {p0={"search.concurrent_segment_search.enabled":"true"}}](https://build.ci.opensearch.org/job/gradle-check/48900/testReport/junit/org.opensearch.search.nested/SimpleNestedIT/testCheckFixedBitSetCache__p0___search_concurrent_segment_search_enabled___true___/)
> [org.opensearch.search.nested.SimpleNestedIT.testCheckFixedBitSetCache {p0={"search.concurrent_segment_search.enabled":"false"}}](https://build.ci.opensearch.org/job/gradle-check/48900/testReport/junit/org.opensearch.search.nested/SimpleNestedIT/testCheckFixedBitSetCache__p0___search_concurrent_segment_search_enabled___false___/)
> ```

Addressed it as part of https://github.com/opensearch-project/OpenSearch/pull/15938/commits/1e78e1e06f77f4932a885de2126d2b84834d3dd5

rajiv-kv commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIndices.java:
```diff
@@ -68,14 +71,58 @@ public class ClusterStatsIndices implements ToXContentFragment {
+    ) {
         Map<String, ShardStats> countsPerIndex = new HashMap<>();
+        this.docs = indicesMetrics.contains(IndexMetric.DOCS) ? new DocsStats() : null;
+        Consumer<DocsStats> docsStatsConsumer = (docs) -> {
```
Consumer can be removed and this can be inline with the for loop

SwethaGuptha replied:
Removed.

rajiv-kv commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIndices.java:
```diff
@@ -113,21 +160,23 @@ public ClusterStatsIndices(List<ClusterStatsNodeResponse> nodeResponses, Mapping
-                    queryCache.add(shardCommonStats.queryCache);
-                    completion.add(shardCommonStats.completion);
-                    segments.add(shardCommonStats.segments);
+                    storeStatsConsumer.accept(shardCommonStats.store);
```
Lets add a test for the case where metrics are requested by user (with filter) but the cluster has no indices.

SwethaGuptha replied:
Added.

SwethaGuptha commented:
> ❌ Gradle check result for [f9ae78c](https://github.com/opensearch-project/OpenSearch/commit/f9ae78cf3ccc6a7b4065c8076fe5cfb8766bdd31): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/48929/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flaky test org.opensearch.cluster.SpecificClusterManagerNodesIT.testElectOnlyBetweenClusterManagerNodes

SwethaGuptha commented:
> ❌ Gradle check result for [c3634ac](https://github.com/opensearch-project/OpenSearch/commit/c3634ac49f74e5b54ed41732031ec87d723808b3): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/48941/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Failed tests:

```
org.opensearch.remotestore.RemoteStorePinnedTimestampsGarbageCollectionIT.testLiveIndexWithPinnedTimestamps
org.opensearch.snapshots.ConcurrentSnapshotsV2IT.testDeleteAndCloneV1WhileV2CreateOngoing
```

SwethaGuptha commented:
> ❌ Gradle check result for [3e3e78c](https://github.com/opensearch-project/OpenSearch/commit/3e3e78cd035cc4732724d3f65fc3e0efbc80f93c): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/48943/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flaky test
https://github.com/opensearch-project/OpenSearch/issues/16199
https://github.com/opensearch-project/OpenSearch/issues/16015

mgodwan commented:
@SwethaGuptha Thanks for the changes.

1. Could you add sample responses for certain filters for better understanding?
2. Could you open documentation issue for the change, and help with the documentation content?

mgodwan commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIndices.java:
```diff
@@ -68,14 +70,23 @@ public class ClusterStatsIndices implements ToXContentFragment {
-        this.queryCache = new QueryCacheStats();
-        this.completion = new CompletionStats();
-        this.segments = new SegmentsStats();
+    public ClusterStatsIndices(
```
Since we deleted the old constructor, we may get backward compatibility check failures in 2.x.
Could you verify the same and ensure we handle it?

SwethaGuptha replied:
Old constructor is still there for this class. Change has been made to call the new constructor from the old constructor with default arguments.

mgodwan commented on server/src/test/java/org/opensearch/action/admin/cluster/stats/ClusterStatsRequestTests.java:
```diff
@@ -0,0 +1,102 @@
+
+import java.util.Set;
+
+public class ClusterStatsRequestTests extends OpenSearchTestCase {
```
We've good tests in `AbstractWireSerializingTestCase` and `AbstractXContentTestCase`.
Can you check if the tests are useful, and we can reuse some of the predefined cases as they already cover a wide variety of Req/Resp ser/deser

SwethaGuptha replied:
Added this test case to specifically test the ser/de od ClusterStateRequest object for versions greater than 3.0 and below 3.0 specifically. I checked the AbstractWireSerializingTestCase, these test case require equals method implementation in the class being tested. This class doesn't have equals method implementation and the class variable types also don't have an equals method implementation. Additionally the test cases in AbstractWireSerializingTestCase only test for current version where as for this change we want to test for versions both below and above 3.0.

shwetathareja replied:
this test has to be updated when merging in 2.x.

SwethaGuptha replied:
Acknowledge.

mgodwan commented on server/src/test/java/org/opensearch/rest/action/admin/cluster/RestClusterStatsActionTests.java:
```diff
@@ -0,0 +1,171 @@
+
+public class RestClusterStatsActionTests extends OpenSearchTestCase {
+
+    private RestClusterStatsAction action;
```
Thanks for adding this test class

mgodwan commented on server/src/main/java/org/opensearch/index/fielddata/FieldDataStats.java:
```diff
@@ -80,6 +80,9 @@ public FieldDataStats(long memorySize, long evictions, @Nullable FieldMemoryStat
     }
 
     public void add(FieldDataStats stats) {
+        if (stats == null) {
```
When can this be null?

SwethaGuptha replied:
It can be null when the client doesn't request fielddata as part of index_metrics.

mgodwan reviewed (changes requested):
Few comments on the testing side:

1. Lets add a rest-api-spec test to ensure the different rest paths are tested?
2. Can we added a mixed cluster test case to ensure rolling upgrades continue to work seamlessly?

SwethaGuptha commented:
> Few comments on the testing side:
> 
>     1. Lets add a rest-api-spec test to ensure the different rest paths are tested?

  I added integration test case RestClusterStatsAction to validate different rest paths, is this some different kind of test case?
>     2. Can we added a mixed cluster test case to ensure rolling upgrades continue to work seamlessly?

Can you please share a reference for this kind of test?

SwethaGuptha commented:
> @SwethaGuptha Thanks for the changes.
> 
>     1. Could you add sample responses for certain filters for better understanding?
Updated PR description with sample responses.

>     2. Could you open documentation issue for the change, and help with the documentation content?

Doc issue: https://github.com/opensearch-project/documentation-website/issues/8527

shwetathareja commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsNodes.java:
```diff
@@ -101,26 +121,29 @@ public class ClusterStatsNodes implements ToXContentFragment {
                 continue;
             }
-            if (nodeResponse.nodeStats().getFs() != null) {
+            if (this.fs != null && nodeResponse.nodeStats().getFs() != null) {
```
same here as well, check for requestedMetric instead of null check on fs

SwethaGuptha commented on server/src/main/java/org/opensearch/action/admin/cluster/stats/ClusterStatsIndices.java:
```diff
@@ -92,12 +103,24 @@ public ClusterStatsIndices(List<ClusterStatsNodeResponse> nodeResponses, Mapping
-                queryCache.add(r.getAggregatedNodeLevelStats().commonStats.queryCache);
-                completion.add(r.getAggregatedNodeLevelStats().commonStats.completion);
-                segments.add(r.getAggregatedNodeLevelStats().commonStats.segments);
+                if (docs != null) {
```
Addressed.

SwethaGuptha commented:
> ❌ Gradle check result for [da52aeb](https://github.com/opensearch-project/OpenSearch/commit/da52aebee360cbd145fee01e5cf21cfb061c8c64): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/49892/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Flaky test: https://github.com/opensearch-project/OpenSearch/issues/15812

shwetathareja commented:
@SwethaGuptha please add changelog for it

shwetathareja commented:
we will need documentation PR and Rest API spec PR as well for this change.

SwethaGuptha commented:
> we will need documentation PR and Rest API spec PR as well for this change.

Raised PR for both of them.
